wolfhsm/wh_packet.h

```diff
@@ -307,6 +307,36 @@ typedef struct WOLFHSM_PACK wh_Packet_version_exchange
     uint32_t version;
 } wh_Packet_version_exchange;
 
+typedef struct WOLFHSM_PACK wh_Packet_counter_reset_req
+{
+    uint16_t keyId;
+} wh_Packet_counter_reset_req;
+
+typedef struct WOLFHSM_PACK wh_Packet_counter_reset_res
+{
+    uint32_t counter;
+} wh_Packet_counter_reset_res;
+
+typedef struct WOLFHSM_PACK wh_Packet_counter_increment_req
+{
+    uint16_t keyId;
+} wh_Packet_counter_increment_req;
+
+typedef struct WOLFHSM_PACK wh_Packet_counter_increment_res
+{
+    uint32_t counter;
+} wh_Packet_counter_increment_res;
+
+typedef struct WOLFHSM_PACK wh_Packet_counter_read_req
+{
+    uint16_t keyId;
+} wh_Packet_counter_read_req;
+
+typedef struct WOLFHSM_PACK wh_Packet_counter_read_res
+{
+    uint32_t counter;
+} wh_Packet_counter_read_res;
+
```

Comment:
hmm, I'm noticing we have a split paradigm developing here - you have been putting messages into the `whPacket` union in this file, like you do for crypto, whereas Bill and I have been adding new message definitions and the necessary marshalling/data transformations in their own `wh_message_xxx.{c,h}` files. I think we should standardize on doing it one way. 

Related, I see there aren't any endian transformations for these messages. This will break if we ever have platforms with mixed HSM and host endianness. The reason Bill and I implemented the intermediate message layer is to allow a distinct module that handles the endian transformations for a given message data type, between host format and HSM format. 

I'm also now realizing we don't do that for any of the crypto...I think that is something we will need to add at some point.

I suppose we could punt it for now, though I'd really prefer adding the appropriate transformation support to this module before merging, and just punting the crypto. And separating this from crypto in its entirety, as it is a completely different feature/module

Reply:
I think everything should be put in packet.h and all packets should be defined in the single unionionized struct that I've been using so we don't have to serialize and deserialize anything and can use a single piece of memory for the packet by just casting. We will need to add networkize and hostize functions that swap the endianness in a single switch statement based on message group and action. frankly adding new files for all this just makes more technical debt

Reply:
Counters should be supported even when crypto is not.  We'll need to move things around in the future, but we'll stay the course for now.  Providing separate translation units means fewer conditional compilation flags, so the giant union isn't always the easiest way to manage bulk data.  Each individual kind of message will be typecast anyway, and should be processed in separate functions, so the union only exists at the main dispatch.

Again, nothing to fix at this point, but when we get a chance, we really should focus on eliminating "packed" by reorganizing the structs and using fixed size types in message fields.  Then, the only processing per field is for endianness.

Reply:
Well I think we will always need to serialize/deserialize. To clarify, when I say serialize/deserialize I'm referring to the same thing as you are when you say networkize/hostize. I suppose "marshalling" is a better word to use. Going from a standard "wire format" to a native host format. So if we want to support platforms with different byte orders, there isn't really any way we can get around this operation. And it goes beyond endianness too, it includes packing/padding of structs, which can be annoyingly different for different compilers/toolchains/architectures.

And the marshaling is already is done based on the message group and action like you are suggesting, but it just is handled one level down (e.g. NVM messages are deserialized in `wh_Server_HandleNvmRequest`, Comm messages are deserialized in `wh_Server_HandleCommRequest`, etc.) instead of before the message dispatch in `wh_Server_HandleRequest()`. Not sure why doing it this way results in more technical debt. Some might say more files mean less code in a given context, making things clearer, and there is more of a clear separation of concerns. Not at all trying to die on this hill, I just don't think it is necessarily a problem. What is a problem IMO, is not doing any proper marshalling of data and manually reordering/padding the structs to try and avoid issues.

That said, I've long been thinking about using a ser/de library (or making our own simple one) so wolfHSM itself doesn't need to worry about this stuff. Something like protobuf/flatbuffers or the like. Since using packed structs on different architectures is pretty dangerous, and right now we are just manually packing/aligning structs which seems like not the best way.

Ideally we have a struct representation of each data type that is a normal native struct able to be used on both sides, and we can just call `myStruct_deSerialize(&myLocalStruct, &commBuffer)` or `myStruct_serialize(&myLocalStruct, &commBuffer)`  and the data would be properly marshalled without worrying about endianness, padding, alignment, etc.

Add it to the backlog, I suppose...